In the X.509 SVID spec, make EKU requirements more flexible.

briansmith opened the pull request:
Leave the door open for future work on EKU restrictions in intermediate
CA certifictes.

An empty EKU field means "all EKUs are allowed." Therefore, it doesn't
make much sense to allow an empty EKU but then forbid EKUs that have
EKUs other than id-kp-serverAuth and id-kp-clientAuth; if other EKUs
are truly harmful to be included then a missing EKU extension should be
forbidden too.

briansmith committed:
In the X.509 SVID spec, make EKU requirements more flexible.

Leave the door open for future work on EKU restrictions in intermediate
CA certifictes.

An empty EKU field means "all EKUs are allowed." Therefore, it doesn't
make much sense to allow an empty EKU but then forbid EKUs that have
EKUs other than id-kp-serverAuth and id-kp-clientAuth; if other EKUs
are truly harmful to be included then a missing EKU extension should be
forbidden too.

rra commented:
This looks reasonable to me, although do you have some specifics on the statement here that EKU restrictions vary widely on intermediate certs?

briansmith commented:
> do you have some specifics on the statement here that EKU restrictions vary widely on intermediate certs?

In Microsoft's implementation of X.509, and in Firefox's (mozilla::pkix) and the Rust [webpki](https://github.com/briansmith/webpki) implementation, any EKU constraints in intermediate CA certificates are enforced on the certificates below. So, for example, if the intermediate CA has only id-kp-serverAuth and the end-entity certificate has both id-kp-serverAuth and id-kp-clientAuth, then they will only trust the certificate for id-kp-serverAuth. However, this isn't universally implemented. I believe that OpenSSL, for example, does things differently.

rra commented:
Thank you!  I think that would be a useful comment to add to the spec itself in some abbreviated form, like: "Some X.509 implementations will impose EKU constraints in intermediate CA certificates on all certificates below them in the trust chain, and others will not."

evan2645 commented on standards/X509-SVID.md:
```diff
@@ -77,9 +77,10 @@ Leaf SVIDs MUST set `keyEncipherment`, `keyAgreement`, and `digitalSignature`. T
 
-Signing certificates MUST NOT include an Extended Key Usage extension.
+Signing certificates SHOULD NOT include an Extended Key Usage extension because the processing of the Extended Key Usage extension
+in intermediate CA certificates varies widely across X.509 certificate validation libraries.
```
Erroneous newline?

evan2645 reviewed (commented):
Agreed with @rra, a "for instance" example would be useful here since we are giving a justification

briansmith committed:
Remove accidentally-added newline.

briansmith committed:
Add additional explaination requested in PR review.

briansmith commented:
I removed the extra newline and added the suggested statement.

evan2645 reviewed (approved):
LGTM, needs one more approver

briansmith committed:
Change "SHOULD NOT" to "MAY" in EKU requirements for signing certs.

"SHOULD NOT" is too strong. Although there *is* variability in the
handling of EKU amongst implementations, in general they either enforce
the EKU for a signing (CA) cert hierarchically or not at all. This
isn't a risk to compatibility or security.

justinburke commented on standards/X509-SVID.md:
```diff
@@ -77,9 +77,9 @@ Leaf SVIDs MUST set `keyEncipherment`, `keyAgreement`, and `digitalSignature`. T
 ### 4.4. Extended Key Usage
 This extension indicates one or more purposes for which the key contained in the certificate may be used, in addition to or in place of the basic purposes indicated in the key usage extension. It is defined in [RFC 5280, section 4.2.1.2][6].
```
I think this should be section 4.2.1.12.

justinburke commented:
LGTM.
